gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -181,12 +184,51 @@ private ManagedChannel createSingleChannel() throws IOException {
     int port = Integer.parseInt(endpoint.substring(colon + 1));
     String serviceAddress = endpoint.substring(0, colon);
 
-    ManagedChannelBuilder builder =
-        ManagedChannelBuilder.forAddress(serviceAddress, port)
-            .intercept(headerInterceptor)
-            .intercept(metadataHandlerInterceptor)
-            .userAgent(headerInterceptor.getUserAgentHeader())
-            .executor(executor);
+    // TODO(hzyi): Use NettyChannelBuilder when maxHeaderListSize is specified to unblock Spanner.
+    // Change to ManagedChannelBuilder when https://github.com/grpc/grpc-java/issues/4050 is
+    // resolved.
+    ManagedChannelBuilder builder;
+    if (maxHeaderListSize != null) {
+      Class<?> nettyChannelBuilderClass;
+      try {
+        nettyChannelBuilderClass =
+            Class.forName("io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder");
+      } catch (ClassNotFoundException e) {
+        try {
+          nettyChannelBuilderClass = Class.forName("io.grpc.netty.NettyChannelBuilder");
+        } catch (ClassNotFoundException ex) {
+          throw new RuntimeException(
+              "Unable to create the channel because neither"
+                  + " \"io.grpc.netty.NettyChannelBuilder\" nor"
+                  + " \"io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder\" is found.");
+        }
+      }
+      try {
+        Object object =
+            nettyChannelBuilderClass
+                .getMethod("forAddress", String.class, int.class)
+                .invoke(null, serviceAddress, port);
+        object =
+            nettyChannelBuilderClass
+                .getMethod("maxHeaderListSize", int.class)
+                .invoke(object, maxHeaderListSize);
```

Comment:
To be more explicit, I think the general structure should be like this:

```
ManagedChannelBuilder builder;
try {
  shadedNettyClass = Class.forName(...);
  shadedNettyBuilder = io...shaded...NettyChannelBuilder.forAddress(...);
  shadedNettyBuilder.maxHeaderListSize(...);
  builder = shadedNettyBuilder;
} catch (CNFE e) {
  try {
    directNettyClass = Class.forName(...);
    directNettyBuilder = io...NettyChannelBuilder.forAddress(...);
    directNettyBuilder.maxHeaderListSize(...);
    builder = directNettyBuilder;
  } catch (CNFE e) {
    throw ...;
  }
}
builder...
```

Maybe you can only do this if you declare a dependency on both `grpc-netty` and `grpc-netty-shaded`; if so, the `grpc-netty` should be a `provided` dependency so that it doesn't show up on consumers' build graph.